### PR TITLE
feat: stream data from request

### PR DESCRIPTION
**Pull request by jdu-mydevices**
https://mydevices.atlassian.net/browse/SS-4946
https://github.com/myDevicesIoT/tina-api/pull/727

### CODE REVIEW

File: lib/Auth/index.js [hunk 1]
```diff
@@ -156,7 +156,7 @@ class Auth {
    *
    * @throws {CoreError} Returns erroror max authenticated request limit reached
    */
-  async send(service, method, url, opts, retry = 0) {
+  async send(service, method, url, opts, retry = 0, stream = false) {
```
Comment (hugocortes) [inline lib/Auth/index.js]: Let's add the new variable to `opts` and initialize the default value within Object.assign

```js
  /**
   * Send a request to Core service
....
   * @param {Object}  [opts.headers={}]     Headers to send
   * @param {Object}  [opts.query]          Query to send
   * @param {Object}  [opts.payload]        Payload to send
  *  @param {Object}  [opts.stream]
...
...
   */
  async send(service, method, url, opts, retry = 0) {
...
    opts = Object.assign(
      {
        ...
        ...
        stream: false
      },
     opts
    )
  }
```

File: lib/Auth/index.js [hunk 2]
```diff
@@ -204,6 +204,19 @@ class Auth {
           opts.headers.authorization = `Bearer ${clientToken}`;
         }
 
+        if (stream) {
```
Comment (hugocortes) [inline lib/Auth/index.js]: We can create a superagent object with the common fields between streaming and no stream request.

```js
const req = superagent(method, url)
  .set(opts.headers)
  .timeout( .... )
  .query(opts.query)
  .send(opts.payload)

if (opts.stream) {
  return req.pipe(opts.stream)
}
begin = new Date().getTime();
({ status, body } = await req());
```

File: lib/Auth/index.js (hunk 1 above)
Comment (jdu-mydevices) [reply lib/Auth/index.js]: Moved stream to opts

File: lib/Auth/index.js (hunk 2 above)
Comment (jdu-mydevices) [reply lib/Auth/index.js]: Moved stream to opts

File: lib/Services/delorean-es.js [hunk 3]
```diff
@@ -106,6 +106,29 @@ class DeloreanES {
...
+  /**
+   * @param {String}  userId
+   * @param {String}  deviceId
+   * @param {Object}  query
+   * @param {Object}  stream
+   */
+  getDeviceReportStream(userId, deviceId, query, stream) {
+    const url = `${this.url}/devices/${deviceId}/stream/report`;
+    query = { ...query, user_id: userId };
+    return this.auth.send(this.service, 'GET', url, { query, stream }, 0);
```
Comment (hugocortes) [inline lib/Services/delorean-es.js]: As response timeout was being increased, I suggest adding an opts variable to method signature and pass that through send function to allow higher timeouts for streams.

And `0` is not required, this is only to be used for authorization failure fallback.

```js
getDeviceReportStream(userId, deviceId, query, opts) {
  // stream, response, and deadline will be passed through and override default timeouts
  return this.auth.send(this.service, 'GET', url, { query, ...opts });
}
```

File: lib/Auth/index.js [hunk 4]
```diff
@@ -204,15 +205,20 @@ class Auth {
...
         }
 
-        begin = new Date().getTime();
-        ({ status, body } = await superagent(method, url)
+        const req = superagent(method, url)
           .set(opts.headers)
           .timeout({
-            response: opts.response,
-            deadline: opts.deadline
+            response: opts.response * 3,
```
[the remark this answers is not in the document]
Comment (jdu-mydevices) [reply lib/Auth/index.js]: Only report stream need a high timeout. ES Stream respond to Tina page by page, ?00k level data is normal for 1 month report. Short large requests will cause ES high load without response. Long small requests can help reduce impact to ES.

File: lib/Services/delorean-es.js (hunk 3 above)
Comment (jdu-mydevices) [reply lib/Services/delorean-es.js]: Only ES stream need a longer timeout. Let's keep this for now and if we need it in another place, let's change it.

File: lib/Auth/index.js (hunk 2 above)
Comment (jdu-mydevices) [reply lib/Auth/index.js]: req isn't a function here. changed to use send  in pipe and await

File: lib/Auth/index.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (hugocortes) [reply lib/Auth/index.js]: If that's the case, let's set this value from `getDeviceReportStream`

```js
return this.auth.send(this.service, 'GET', url, { query, stream, response: 90000*3, timeout: 90000* 3});
```

This overrides the default only for stream and not for every other request.

**Pull request merged**